x/nft/internal/keeper/keeper.go

```diff
@@ -0,0 +1,300 @@
+package keeper
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/nft/internal/types"
+)
+
+// Keeper maintains the link to data storage and exposes getter/setter methods for the various parts of the state machine
+type Keeper struct {
+	storeKey sdk.StoreKey // Unexposed key to access store from sdk.Context
+
+	cdc *codec.Codec // The wire codec for binary encoding/decoding.
+}
+
+// NewKeeper creates new instances of the nft Keeper
+func NewKeeper(cdc *codec.Codec, storeKey sdk.StoreKey) Keeper {
+	return Keeper{
+		storeKey: storeKey,
+		cdc:      cdc,
+	}
+}
+
+// IsNFT returns whether an NFT exists
+func (k Keeper) IsNFT(ctx sdk.Context, denom, id string) (exists bool) {
+	_, err := k.GetNFT(ctx, denom, id)
+	return err == nil
+}
+
+// GetNFT gets the entire NFT metadata struct for a uint64
+func (k Keeper) GetNFT(ctx sdk.Context, denom, id string) (nft types.NFT, err sdk.Error) {
+	collection, found := k.GetCollection(ctx, denom)
+	if !found {
+		return nil, types.ErrUnknownCollection(types.DefaultCodespace, fmt.Sprintf("collection of %s doesn't exist", denom))
+	}
+	nft, err = collection.GetNFT(id)
+
+	if err != nil {
+		return nil, err
+	}
+	return nft, err
+}
+
+// UpdateNFT updates an already existing NFTs
+func (k Keeper) UpdateNFT(ctx sdk.Context, denom string, nft types.NFT) (err sdk.Error) {
+	collection, found := k.GetCollection(ctx, denom)
+	if !found {
+		return types.ErrUnknownCollection(types.DefaultCodespace,
+			fmt.Sprintf("Collection #%s doesn't exist", denom),
+		)
+	}
+	oldNFT, err := collection.GetNFT(nft.GetID())
+	if err != nil {
+		return err
+	}
+	// if the owner changed then update the owners KVStore too
+	if !oldNFT.GetOwner().Equals(nft.GetOwner()) {
+		err = k.SwapOwners(ctx, denom, nft.GetID(), oldNFT.GetOwner(), nft.GetOwner())
+		if err != nil {
+			return err
+		}
+	}
+	collection, err = collection.UpdateNFT(nft)
+	if err != nil {
+		return err
+	}
+	k.SetCollection(ctx, denom, collection)
+	return nil
+}
+
+// MintNFT mints an NFT and manages that NFTs existence within Collections and Owners
+func (k Keeper) MintNFT(ctx sdk.Context, denom string, nft types.NFT) (err sdk.Error) {
```

Comment:
Shouldn't Mint and Burn NFT's be permissioned. Why is everyone allowed to mint and burn NFTs?

Reply:
Great question! We thought it would be helpful to provide these sorts of msg types to all users however when actually using the module it should be advised to overwrite the handler with custom logic. This will allow an NFT to be selective about who can mint/burn and when. It also allows them to discard those msg types entirely if they have no place in their use. You can see an example of this usage in the sample NFT app here: 
https://github.com/okwme/cosmos-nft/blob/master/app.go#L196
and the handler functions here:
https://github.com/okwme/cosmos-nft/blob/master/handler.go